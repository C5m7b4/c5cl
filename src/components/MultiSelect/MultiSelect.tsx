import React, { useState, useEffect } from 'react';
import MultiSelectPortal from './MultiSelectPortal';

import './MultiSelect.css';

export interface MultiSelectProps<T> {
  data: T[];
  defaultOption: string;
  keyDescriptor: string;
  optionValue: keyof T;
  optionText: keyof T;
  checkedItems: string[];
  setCheckedItems: (s: string[]) => void;
  id?: string;
  selectBoxId?: string;
  backgroundColor?: string;
  selectId?: string;
  inputId?: string;
}

function MultiSelect<T>(props: MultiSelectProps<T>) {
  const [show, setShow] = useState(false);
  const [filteredData, setFilteredData] = useState<T[]>([]);
  const [search, setSearch] = useState('');
  const [locallyCheckedItems, setLocallyCheckedItems] = useState<string[]>([]);

  const {
    id = 'multi-select-checkboxes',
    selectBoxId = 'select-box-id',
    backgroundColor = '#1f1e2e',
    selectId = 'select-id',
    inputId = 'multi-select-search',
    checkedItems,
    setCheckedItems,
    optionValue,
    keyDescriptor,
    optionText,
  } = props;

  useEffect(() => {
    setFilteredData(props.data);
    setLocallyCheckedItems(checkedItems);
  }, [props.data]);

  const amIChecked = (i: T) => {
    if (locallyCheckedItems.includes(i[optionValue] as unknown as string)) {
      return true;
    } else {
      return false;
    }
  };

  const tapOutside = (e: MouseEvent) => {
    const portal = document.getElementById(id);
    if (portal) {
      const rect = portal.getBoundingClientRect();
      /* istanbul ignore else */
      if (e.clientX < rect.left || e.clientX > rect.right) {
        setShow(false);
        // const actualSelect = document.getElementById(selectId);
        // const option = new Option(props.defaultOption, '0');
        // actualSelect?.prepend(option);
        window.removeEventListener('click', tapOutside);
        return;
      }
      /* istanbul ignore else */
      if (e.clientY < rect.top || e.clientY > rect.bottom) {
        setShow(false);
        // const actualSelect = document.getElementById(selectId);
        // const option = new Option(props.defaultOption, '0');
        // actualSelect?.prepend(option);
        window.removeEventListener('click', tapOutside);
        return;
      }
    } else {
      window.removeEventListener('click', tapOutside);
    }
  };

  const showCheckboxes = () => {
    if (show) {
      setShow(false);
      return;
    } else {
      setShow(true);
    }

    const actualSelect = document.getElementById(selectId) as HTMLSelectElement;
    const firstChild = actualSelect.childNodes[0] as HTMLDivElement;
    /* istanbul ignore else */
    if (firstChild) {
      firstChild.style.display = 'none';
    }
    /* istanbul ignore else */
    if (actualSelect) {
      //actualSelect.options.remove(0);

      setTimeout(() => {
        const select = document.getElementById(selectBoxId);
        const portal = document.getElementById(id);
        /* istanbul ignore else */
        if (select && portal) {
          const portalRect = portal.getBoundingClientRect();
          const selectRect = select.getBoundingClientRect();
          portal.style.top = selectRect.bottom + 'px';
          portal.style.left = selectRect.left + 'px';
          portal.style.backgroundColor = backgroundColor;
          window.addEventListener('click', tapOutside);
          const input = document.getElementById(inputId);
          /* istanbul ignore else */
          if (input) {
            input.focus();
          }
        }
      }, 100);
    }
  };

  const handleCheck = (e: React.ChangeEvent<HTMLInputElement>, i: T) => {
    const columnValue = i[optionValue] as unknown as string;

    if (e.target.checked) {
      /* istanbul ignore else */
      if (!locallyCheckedItems.includes(columnValue)) {
        setCheckedItems([...checkedItems, columnValue]);
        setLocallyCheckedItems([...locallyCheckedItems, columnValue]);
      }
    } else {
      const newArray = locallyCheckedItems.filter((i) => i != columnValue);
      setCheckedItems(newArray);
      setLocallyCheckedItems(newArray);
    }
  };

  const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value);
    if (e.target.value.length === 0) {
      setFilteredData(props.data);
    } else {
      const filtered = props.data.filter((s) => {
        const recordValue = s[optionText] as unknown as string;
        /* istanbul ignore else */
        if (
          recordValue.toLowerCase().startsWith(e.target.value.toLowerCase())
        ) {
          return s;
        }
      });
      setFilteredData(filtered);
    }
  };

  return (
    <div className="multi-select">
      <div
        className="select-box"
        id={selectBoxId}
        onClick={() => showCheckboxes()}
      >
        <select className="form-control" id={selectId}>
          <option value="0">{props.defaultOption}</option>
        </select>
      </div>
      <div id="ms-portal-container"></div>
      <MultiSelectPortal
        open={show}
        data={filteredData}
        id={id}
        optionValue={optionValue}
        keyDescriptor={keyDescriptor}
        optionText={optionText}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        search={search}
        amIChecked={amIChecked}
        inputId={inputId}
      />
    </div>
  );
}

export default MultiSelect;
