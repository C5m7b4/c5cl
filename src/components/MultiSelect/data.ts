export const data = [
  {
    id: 1,
    value: 'Grocery',
  },
  {
    id: 2,
    value: 'Nonfood',
  },
  {
    id: 3,
    value: 'HBA',
  },
  {
    id: 4,
    value: 'Meat',
  },
  {
    id: 5,
    value: 'Meat by #',
  },
  {
    id: 6,
    value: 'Produce',
  },
  {
    id: 7,
    value: 'Produce by #',
  },
];
