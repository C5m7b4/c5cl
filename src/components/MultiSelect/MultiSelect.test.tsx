import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import MultiSelect from './MultiSelect';
import { data } from './data';

describe('MultiSelect', () => {
  test('should render correctly', () => {
    const setCheckedItems = jest.fn();
    jest.useFakeTimers();
    jest.spyOn(global, 'setTimeout');
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
    expect(setTimeout).toHaveBeenCalled();
  });
  test('should render without some default props', () => {
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    expect(container).toMatchSnapshot();
    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    fireEvent.click(option);
  });
  test('should handle click outside', () => {
    const div = document.createElement('div');
    div.id = 'outside';
    document.body.appendChild(div);
    const portal = document.createElement('div');
    portal.id = 'ms-portal-container';
    document.body.appendChild(portal);
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);

    const outside = document.querySelector('outside');
    if (outside) {
      fireEvent.click(outside);
    }
  });
  test('should handle check', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );

    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);

    const dept = getByRole('checkbox', { name: /Grocery/ });
    fireEvent.click(dept);
    fireEvent.click(dept);
    fireEvent.click(dept);
  });
  test('should handle search', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);
    const setCheckedItems = jest.fn();
    const { container, getByRole, getByPlaceholderText } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptor="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
      />
    );
    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);

    userEvent.type(screen.getByPlaceholderText(/search/i), 'gro');
    const input = getByPlaceholderText('search') as HTMLInputElement;
    expect(input.value).toEqual('gro');

    userEvent.type(
      screen.getByPlaceholderText(/search/i),
      '{backspace}{backspace}{backspace}'
    );
    expect(input.value).toEqual('');
  });
});
