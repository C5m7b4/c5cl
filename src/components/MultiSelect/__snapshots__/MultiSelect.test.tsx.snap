// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect should render correctly 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="selectBoxId"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;

exports[`MultiSelect should render without some default props 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="select-box-id"
    >
      <select
        class="form-control"
        id="select-id"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;
